lib/datasets.go

```diff
@@ -1736,22 +1179,534 @@ type datasetImpl struct{}
+		// dataset name
+		if p.Outfile == "" && p.GenFilename {
+			p.Outfile = fmt.Sprintf("%s.zip", ds.Name)
+		}
+		var outBuf bytes.Buffer
+		var zipFile io.Writer
+		if p.Outfile == "" {
+			// In this case, write to a buffer, which will be assigned to res.Bytes later on
+			zipFile = &outBuf
+		} else {
+			zipFile, err = os.Create(p.Outfile)
+			if err != nil {
+				return nil, err
+			}
+		}
+		currRef := dsref.Ref{Username: ds.Peername, Name: ds.Name}
+		// TODO(dustmop): This function is inefficient and a poor use of logbook, but it's
+		// necessary until dscache is in use.
+		initID, err := scope.Logbook().RefToInitID(currRef)
+		if err != nil {
+			return nil, err
+		}
+		err = archive.WriteZip(scope.Context(), scope.Filesystem(), ds, "json", initID, currRef, zipFile)
+		if err != nil {
+			return nil, err
+		}
+		// Handle output. If outfile is empty, return the raw bytes. Otherwise provide a helpful
+		// message for the user
+		if p.Outfile == "" {
+			res.Bytes = outBuf.Bytes()
+		} else {
+			res.Message = fmt.Sprintf("Wrote archive %s", p.Outfile)
+		}
+		return res, nil
+	}
+
+	if p.Selector == "body" {
+		// `qri get body` loads the body
+		if !p.All && (p.Limit < 0 || p.Offset < 0) {
+			return nil, fmt.Errorf("invalid limit / offset settings")
+		}
+		df, err := dataset.ParseDataFormatString(p.Format)
+		if err != nil {
+			log.Debugf("Get dataset, ParseDataFormatString %q failed, error: %s", p.Format, err)
+			return nil, err
+		}
+
+		if fsi.IsFSIPath(ref.Path) {
+			// TODO(dustmop): Need to handle the special case where an FSI directory has a body
+			// but no structure, which should infer a schema in order to read the body. Once that
+			// works we can remove the fsi.GetBody call and just use base.ReadBody.
+			res.Bytes, err = fsi.GetBody(fsi.FilesystemPathToLocal(ref.Path), df, p.FormatConfig, p.Offset, p.Limit, p.All)
+			if err != nil {
+				log.Debugf("Get dataset, fsi.GetBody %q failed, error: %s", res.FSIPath, err)
+				return nil, err
+			}
+			err = maybeWriteOutfile(p, res)
+			if err != nil {
+				return nil, err
+			}
+			return res, nil
+		}
+		res.Bytes, err = base.ReadBody(ds, df, p.FormatConfig, p.Limit, p.Offset, p.All)
+		if err != nil {
+			log.Debugf("Get dataset, base.ReadBody %q failed, error: %s", ds, err)
+			return nil, err
+		}
+		err = maybeWriteOutfile(p, res)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	} else if scriptFile, ok := scriptFileSelection(ds, p.Selector); ok {
+		// Fields that have qfs.File types should be read and returned
+		res.Bytes, err = ioutil.ReadAll(scriptFile)
+		if err != nil {
+			return nil, err
+		}
+		err = maybeWriteOutfile(p, res)
+		if err != nil {
+			return nil, err
+		}
+		return res, nil
+	} else if p.Selector == "stats" {
+		statsParams := &StatsParams{
+			Dataset: res.Dataset,
+		}
+		sa, err := scope.Loader().Dataset().Stats(scope.Context(), statsParams)
```

Comment:
No need to get the Instance and go through another `Dispatch` invocation here. All `lib.Stats` is doing is resolving and loading the dataset, but we've already done that by this point.

Instead, add a method `Stats` to scope:

```
func (s *scope) Stats() *stats.Service {
  return s.inst.stats
}
```

and call that, passing in `res.Dataset`.